### PR TITLE
Re-workaround uncancellable `Console.ReadKey`

### PR DESCRIPTION
**Pull request by andschwa**
First pulled out the old native `UnixConsoleEcho` and I think finished #1745, then removed all the `ReadKeyAsync` code that was unreferenced. Currently testing using the `KeyAvailable` workaround on Windows, and testing the buffered `ReadKey` workaround on macOS. They're both broken in different ways 🤷

### CODE REVIEW

**Review by SeeminglyScience — APPROVED** [review]
LGTM, this is just removing dead code and stripping out UnixConsoleEcho yeah?

**Comment by andschwa** [comment]
> LGTM, this is just removing dead code and stripping out UnixConsoleEcho yeah?

So far, yeah.

**Comment by SeeminglyScience** [comment]
Oh right WIP *OOPS* 😁

**Review by SeeminglyScience — APPROVED** [review]
LGTM! Just a couple of small things.

Well done, this is super exciting :heart:

Just going to approve, if you have auto merge on nbd

File: src/PowerShellEditorServices/Services/PowerShell/Host/PsesInternalHost.cs [hunk 1]
```diff
@@ -869,7 +882,16 @@ private ConsoleKeyInfo ReadKey(bool intercept)
...
 
-            _lastKey = ConsoleProxy.SafeReadKey(intercept, CancellationToken.None);
+            // NOTE: This requests that the client (the Code extension) send a non-printing key back
+            // to the terminal on stdin, emulating a user pressing a button. This allows
+            // PSReadLine's thread waiting on Console.ReadKey to return. Normally we'd just cancel
+            // this call, but the .NET API ReadKey is not cancellable, and is stuck until we send
+            // input. This leads to a myriad of problems, but we circumvent them by pretending to
+            // press a key.
+            using CancellationTokenRegistration registration = _readKeyCancellationToken.Register(
+                () => _languageServer?.SendNotification("powerShell/sendKeyPress")
```
[the remark this answers is not in the document]
Comment (dkattan) [reply src/PowerShellEditorServices/Services/PowerShell/Host/PsesInternalHost.cs:892]: Dumb guy question. If we can tell the client to send a keypress, why aren't we letting the client send us keypresses?

File: src/PowerShellEditorServices/Services/PowerShell/Console/ConsoleProxy.cs [hunk 2]
```diff
@@ -181,19 +40,8 @@ static ConsoleProxy()
...
-            catch (OperationCanceledException)
-            {
-                return new ConsoleKeyInfo(
-                    keyChar: ' ',
-                    ConsoleKey.DownArrow,
-                    shift: false,
-                    alt: false,
-                    control: false);
-            }
+            ConsoleKeyInfo key = System.Console.ReadKey(intercept);
```
[the remark this answers is not in the document]
Comment (andschwa) [reply src/PowerShellEditorServices/Services/PowerShell/Console/ConsoleProxy.cs:43]: Unfortunately we're in `namespace Microsoft.PowerShell.EditorServices.Services.PowerShell.Console` so it needs `System` to differentiate it.

File: src/PowerShellEditorServices/Services/PowerShell/Console/ConsoleProxy.cs [hunk 3]
```diff
@@ -181,19 +40,8 @@ static ConsoleProxy()
...
+            return cancellationToken.IsCancellationRequested ? s_nullKeyInfo : key;
```
[the remark this answers is not in the document]
Comment (andschwa) [reply src/PowerShellEditorServices/Services/PowerShell/Console/ConsoleProxy.cs:44]: I double checked that PSReadLine wasn't depending on this or anything, and it doesn't look like it to me: https://github.com/PowerShell/PSReadLine/blob/dc38b451bee4bdf07f7200026be02516807faa09/PSReadLine/CharMap.cs#L153-L156 I think it was a holdover from old code.

File: src/PowerShellEditorServices/Services/PowerShell/Console/ConsoleProxy.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (SeeminglyScience) [reply src/PowerShellEditorServices/Services/PowerShell/Console/ConsoleProxy.cs:43]: AH that's right thanks

**Review by SeeminglyScience — APPROVED** [review]
LGTM!

**Comment by andschwa** [comment]
Yay! Thanks @SeeminglyScience. Merging.

**Comment by andschwa** [comment]
I _believe_ (through my own testing) that this PR fixed:

- https://github.com/PowerShell/vscode-powershell/issues/3881
- https://github.com/PowerShell/vscode-powershell/issues/3756
- https://github.com/PowerShell/vscode-powershell/issues/2741
- https://github.com/PowerShell/vscode-powershell/issues/3876
- https://github.com/PowerShell/vscode-powershell/issues/3225

and certainly others!

File: src/PowerShellEditorServices/Services/PowerShell/Host/PsesInternalHost.cs [hunk 4]
```diff
@@ -862,23 +875,53 @@ private void OnCancelKeyPress(object sender, ConsoleCancelEventArgs args)
...
-            _lastKey = ConsoleProxy.SafeReadKey(intercept, CancellationToken.None);
-            return _lastKey.Value;
+            // NOTE: This requests that the client (the Code extension) send a non-printing key back
+            // to the terminal on stdin, emulating a user pressing a button. This allows
+            // PSReadLine's thread waiting on Console.ReadKey to return. Normally we'd just cancel
+            // this call, but the .NET API ReadKey is not cancellable, and is stuck until we send
+            // input. This leads to a myriad of problems, but we circumvent them by pretending to
+            // press a key, thus allowing ReadKey to return, and us to ignore it.
+            using CancellationTokenRegistration registration = _readKeyCancellationToken.Register(
+                () => _languageServer?.SendNotification("powerShell/sendKeyPress"));
```
Comment (daxian-dbw) [inline src/PowerShellEditorServices/Services/PowerShell/Host/PsesInternalHost.cs:899]: @SeeminglyScience I chatted with @andschwa about this line of code, and wonder will this cause multiple delegates that do the same thing gets registered to the token, which may end up sending multiple keys to terminal when the token is cancelled?
Would it be better to do this before passing the token to `ReadLine` method?

Comment (SeeminglyScience) [reply src/PowerShellEditorServices/Services/PowerShell/Host/PsesInternalHost.cs:899]: The disposal (from `using`) after `ReadKey` *should* take care of that for us.

I worry that if we register it before `ReadKey` is actually called then we're setting yourselves up for a race condition. If it's registered at `ReadLine` and the cancellation happens before `ReadKey` is actually called, we'll end up successfully cancelling *and* also sending input that we will no longer be able to capture and throw away.

Comment (andschwa) [reply src/PowerShellEditorServices/Services/PowerShell/Host/PsesInternalHost.cs:899]: That was my understanding as well.

Comment (daxian-dbw) [reply src/PowerShellEditorServices/Services/PowerShell/Host/PsesInternalHost.cs:899]: Thanks for the clarification. That makes sense!

**Pull request merged**